### ISSUE TITLE
how2use.py returns different embedding in different runs

### ISSUE THREAD

**Issue by BinMl** [palmtreemodel/PalmTree#5]
Thanks for the great efforts of developing PalmTree. I have a shallow question about the example code `how2use.py`.

I download the pre-trained model and run how2use.py twice. Different embeddings of the same example block are returned. May I know whether it is a normal behavior?

For example, the returned embeddings for the last instruction is 

```python
[ 1.81133199e+00 -3.23746109e+00  1.71083677e+00  3.38666618e-01
  -3.75499040e-01 -1.32175195e+00 -1.38709199e+00 -4.27015543e+00
  -1.09782541e+00  2.22386146e+00  2.11417437e-01 -4.84449720e+00
   5.18921623e-03  1.76847649e+00 -8.63245606e-01 -2.69446135e-01
   5.15590966e-01 -4.68213320e+00  1.59906518e+00  1.36549664e+00
   1.88308871e+00  1.62097037e-01 -2.21299958e+00 -8.82814705e-01
   2.07408404e+00 -1.18871832e+00 -1.56577575e+00  3.74550366e+00
   3.88258219e+00 -3.41718554e-01  1.91557467e+00  2.39120531e+00
   1.07569742e+00 -2.28171086e+00  1.17116559e+00  1.13075769e+00
  -1.40898442e+00  2.61307502e+00 -9.96037185e-01  1.08197117e+00
   2.82093835e+00  2.81637698e-01  6.87107146e-01 -3.97220850e+00
   2.93721175e+00  1.36344850e+00  2.19264552e-01 -1.10414588e+00
  -9.43879783e-01 -2.97416019e+00 -1.17001665e+00  1.25808132e+00
  -8.65609467e-01 -2.74332255e-01  1.66285384e+00  2.38418198e+00
  -2.82275796e+00  4.97404486e-01 -1.20886791e+00 -5.47487251e-02
   1.16453342e-01  3.65588546e+00 -4.56950635e-01 -1.25120628e+00
   1.69303071e+00  4.91060503e-02 -2.42631936e+00  5.09205341e+00
   1.32799685e+00 -4.65154260e-01 -2.05004573e+00 -1.49539649e+00
  -1.35405958e+00  6.10847734e-02  1.04950130e+00  1.24309349e+00
   2.61679387e+00 -2.90876579e+00  1.88654447e+00  1.33350536e-01
  -1.41114962e+00 -2.93596339e+00 -5.63511372e-01 -1.54529345e+00
   8.81065011e-01  4.12528366e-01  1.65922761e+00 -1.31469023e+00
   1.21055377e+00  7.06855893e-01 -2.37969708e+00 -3.56634307e+00
   4.01415646e-01  1.50992894e+00 -7.92103887e-01 -2.27579951e+00
   2.02454782e+00  1.18772221e+00  1.77974033e+00  5.64101338e-01
   5.76624945e-02  2.53443909e+00 -1.36051333e+00 -7.48676896e-01
  -5.55248380e-01 -1.69060898e+00 -6.97114110e-01  5.28021038e-01
   3.52969623e+00  9.04885292e-01  2.58693790e+00  6.89522386e-01
   1.49169266e+00 -1.48478553e-01  6.68267846e-01  1.10466444e+00
  -3.16179419e+00  2.49033427e+00 -1.04603849e-01  1.61363566e+00
  -1.09224415e+00 -1.38813481e-01  2.22357059e+00 -6.92139149e-01
  -4.31032944e+00  9.37082231e-01 -8.17471325e-01 -3.83139104e-01]
```

and 

```python
[ 2.28092790e+00 -4.27574492e+00  1.64888728e+00  2.12926388e-01
  -7.51085818e-01  7.58945882e-01 -1.31212890e+00 -5.11319304e+00
  -1.58667254e+00  2.47101641e+00 -1.92863390e-01 -4.74655771e+00
   9.53466356e-01  1.13818574e+00 -5.95402300e-01 -8.15724790e-01
   1.15052819e+00 -4.84964037e+00  5.27293324e-01  2.17294621e+00
   3.45318961e+00 -4.31100518e-01 -1.88414037e+00 -1.10849097e-01
   7.34654546e-01 -9.00522530e-01 -2.50598025e+00  4.13120270e+00
   3.54104233e+00 -1.66144401e-01  1.81455886e+00  2.42314720e+00
   6.96992576e-01 -3.27784085e+00  2.00603938e+00  2.30833578e+00
  -1.18637943e+00  2.07157969e+00 -1.61450732e+00  1.18069160e+00
   3.44975591e+00 -6.35813415e-01  4.16395664e-02 -4.48357248e+00
   1.60392189e+00 -2.00171664e-01  2.68193662e-01 -3.36856055e+00
  -2.30127081e-01 -3.40497971e+00 -1.45984614e+00  8.65208328e-01
  -7.21970677e-01  4.98732090e-01  1.38946593e+00  2.40883255e+00
  -2.52281666e+00  4.38466072e-01 -5.59960842e-01 -1.12331951e+00
  -1.10853660e+00  3.04281425e+00  2.45987684e-01  9.18105066e-01
   2.19591093e+00  4.02511895e-01 -2.96253991e+00  4.78521442e+00
   1.79124379e+00 -7.23273337e-01 -2.88906455e+00 -1.15492332e+00
  -8.49803925e-01  3.95330638e-01  8.18893611e-01  1.02965140e+00
   2.45947266e+00 -2.71628261e+00  3.50821987e-02 -3.89349133e-01
  -1.92537057e+00 -3.03571773e+00 -1.51975656e+00 -2.53654289e+00
   1.12211633e+00  6.63661122e-01  1.25590265e+00 -2.15385294e+00
   1.18132734e+00  2.04770323e-02 -2.98344660e+00 -4.25735426e+00
  -1.00465767e-01  5.31164408e-01 -4.06229109e-01 -2.09186625e+00
   2.25581336e+00 -1.44944072e-01  2.24776793e+00  1.01454723e+00
   1.25944793e+00  1.27758813e+00 -1.00158560e+00  7.73148477e-01
  -6.22425340e-02 -1.27718377e+00 -1.88933623e+00  9.10866559e-01
   2.55852723e+00  3.58365566e-01  2.00512195e+00  7.65073240e-01
   1.94424999e+00  6.88421786e-01  1.05114810e-01  1.92407382e+00
  -3.90151763e+00  2.39114380e+00  8.12441856e-02  1.62225688e+00
  -2.75948793e-01 -5.59443653e-01  1.83524787e+00 -1.57599103e+00
  -5.58957911e+00  7.51713693e-01 -3.57316375e-01 -3.29680771e-01]
```

The cosine similarity is 0.9296 which may be insignificant. But if possible, may I know whether it is a normal behavior?

Besides, I am also wondering whether it is possible to feed in a sequence of instructions to PalmTree in a single array (along with the segment array) instead of splitting them into different arrays?

That's said, in `eval_utils.py`, for a given instruction sequence 

```asm
push rbp
mov rbp, rsp
sub rsp, 0x20
```

We create a sequence tensor and a segment tensor as 

```python
[
 [<CLE> push rbp <SEP> <PAD> ...]
 [<CLE> mov rbp rsp <SEP> <PAD> ...]
 [<CLE> sub rsp 0x20 <SEP> <PAD> ...]
]
```

and 

```python
[
 [1 1 1 1 0 0 0 ...]
 [1 1 1 1 1 0 0 0 ...]
 [1 1 1 1 1 0 0 0 ...]
]
```

But can we simply create a single-dimension tensor

```python
[<CLE> push rbp <SEP> mov rbp rsp <SEP> sub rsp 0x20 <SEP> <PAD> ...]
```

along with

```python
[1 1 1 1 2 2 2 2 3 3 3 3 0 0 0 ...]
```

**Comment by rmallof** [comment]
Was how2use.py removed from the repo?

**Comment by bobbyhaliwela** [comment]
> Was how2use.py removed from the repo?

It seems that way. I'm not sure why but maybe @SidekickLx plan to complete it?

**Comment by SidekickLx** [comment]
@rmallof how2use.py is not in the repo. it is in the google drive link in README.
@BinMl It is not expected I think. Because the model has been fixed. So there might be something wrong.
For the second question, PalmTree is an embedding model for single instruction as designed. So the code was written to receive single instruction once. Moreover, different segment tokens [1 1 1 2 2 2 3 3 3] might lead to different result.

**Comment by Adalsteinnjons** [comment]
I also get different results when running how2use.py twice.
Here are the instruction embeddings for the last instruction:
```
[1.93552661e+00 -3.39699936e+00  1.59303606e+00  1.62248522e-01
   4.34285581e-01 -1.34500682e+00 -1.56839561e+00 -5.09599495e+00
  -1.88629746e+00  3.33484697e+00 -5.54149151e-02 -3.32852936e+00
   2.67769694e-02  2.99965930e+00 -1.32540631e+00  5.92380881e-01
   1.34754825e+00 -5.51757812e+00  1.60903096e+00  1.79028928e+00
   2.03798604e+00  8.90257180e-01 -2.75775862e+00 -6.40308499e-01
   2.91920018e+00 -2.46261716e-01 -2.12803340e+00  3.71118402e+00
   3.29799032e+00 -1.07317731e-01  2.21493793e+00  2.69728518e+00
   2.53049326e+00 -1.60039973e+00  9.00557160e-01  1.90734506e+00
   4.40661430e-01  2.02584958e+00 -6.54008865e-01  1.22563708e+00
   2.25516176e+00 -1.27577353e+00 -2.12018013e-01 -4.33074951e+00
   2.14706755e+00  1.62046283e-01 -8.14670920e-01 -1.57485414e+00
   1.26501650e-01 -3.94446826e+00 -1.23732686e+00  6.77162290e-01
  -9.04975295e-01 -8.63390565e-01  1.54483330e+00  1.68089902e+00
  -3.19658208e+00  9.34664369e-01 -1.87099421e+00  7.09801689e-02
  -4.77199793e-01  3.01570630e+00  2.27816671e-01 -2.91870165e+00
   3.53411365e+00  2.61483580e-01 -2.42374849e+00  6.35543823e+00
   4.63319063e-01 -3.28873098e-01 -1.34340596e+00 -1.09483790e+00
  -1.92844677e+00 -8.41282979e-02  1.95847011e+00  3.29157293e-01
   2.41936803e+00 -2.75142765e+00  1.10795188e+00  4.46029246e-01
  -1.91901469e+00 -2.34779906e+00 -3.62781852e-01 -1.81776619e+00
  -6.48574412e-01  2.07154304e-01  2.01782393e+00 -1.03721929e+00
   2.23989511e+00 -3.87877449e-02 -1.88930893e+00 -3.24641037e+00
  -8.60100925e-01  9.37420726e-01 -5.56488812e-01 -1.68712902e+00
   1.73470879e+00 -5.10114193e-01  1.57944775e+00  2.67238021e+00
   3.28169882e-01  1.57226765e+00 -2.32458973e+00 -6.76211536e-01
  -2.81786203e-01 -9.84425426e-01 -1.45796096e+00  8.60794246e-01
   4.34354305e+00  1.33752966e+00  2.23522925e+00  8.84349704e-01
   3.21282816e+00  2.87379801e-01  1.15443015e+00  4.53781597e-02
  -3.65583467e+00  3.43089724e+00 -4.42693532e-01  8.08973312e-01
  -1.18780541e+00 -1.14222550e+00  2.30953407e+00 -7.75040209e-01
  -4.48628092e+00  1.31510901e+00  1.55657260e-02 -7.55608827e-02]
```

And when running it again I get: 
```
[2.12508917e+00 -3.89714003e+00  2.28891826e+00  1.29072487e-01
  -7.61487126e-01 -5.31887472e-01 -2.01745701e+00 -4.95254755e+00
  -1.18915403e+00  2.84221148e+00 -3.81252766e-01 -4.53930473e+00
  -9.76838410e-01  2.72660875e+00 -8.58489990e-01 -8.87019992e-01
   7.48951077e-01 -4.10279942e+00  1.10650218e+00  1.67020106e+00
   1.88143289e+00  3.73853117e-01 -1.85768127e+00  5.44254065e-01
   2.31170130e+00 -1.03076780e+00 -2.03853035e+00  4.17776871e+00
   3.56699610e+00 -1.74747974e-01  1.49788868e+00  2.01480937e+00
   1.43520153e+00 -3.98347855e+00  1.34475112e+00  1.47012973e+00
  -7.96976328e-01  2.42370629e+00 -5.20594895e-01  1.00346780e+00
   1.97939241e+00 -9.38571155e-01  1.03360437e-01 -5.39951992e+00
   3.36114550e+00  8.02514076e-01 -1.77196667e-01 -2.90585279e+00
  -1.21421412e-01 -3.86146164e+00 -1.66525197e+00  1.07692754e+00
  -9.88567352e-01 -3.52869570e-01  2.09216952e+00  2.92349577e+00
  -2.31005359e+00  1.49999714e+00 -3.63559067e-01 -1.07944655e+00
  -1.32698774e+00  3.46247244e+00  6.50525689e-01 -6.75974369e-01
   3.07306242e+00  1.70999438e-01 -7.30359852e-01  5.06733751e+00
   1.83093679e+00 -3.90978754e-01 -1.99864578e+00 -1.48761773e+00
  -1.85081351e+00 -5.74549735e-01  8.60717148e-02  1.58861041e+00
   2.48434734e+00 -2.87159777e+00  7.24741638e-01 -5.91942251e-01
  -1.35690665e+00 -2.28800941e+00 -1.20394659e+00 -1.80960846e+00
  -6.08711243e-02  1.23349977e+00  1.65374064e+00 -3.07325506e+00
   1.03505611e+00  9.04900193e-01 -2.96360111e+00 -3.62171984e+00
  -8.49183857e-01  1.00546682e+00  7.13943422e-01 -2.35664797e+00
   2.21257305e+00 -7.16682434e-01  2.61896753e+00  6.65669799e-01
   1.07716084e+00  2.82584453e+00 -2.12765884e+00  5.18370628e-01
  -6.15814984e-01 -9.71785426e-01 -3.47235489e+00  2.30470419e-01
   3.25739598e+00  8.76015544e-01  3.11718035e+00  3.58450979e-01
   2.64322376e+00  2.61192232e-01  5.59520841e-01  1.99210238e+00
  -4.02838802e+00  2.91891623e+00 -2.17249826e-01  1.52961242e+00
  -1.48679185e+00 -1.40883029e+00  1.12225652e+00 -2.18742818e-01
  -5.32906532e+00  6.73483491e-01 -6.73031271e-01 -6.34712100e-01]
```

**Comment by SidekickLx** [comment]
Not an expected result.
Need to check the code.

**Comment by 0599jiangyc** [comment]
I also meet this problem. ```How2use.py``` gives different embeddings every time :(

**Comment by tyj1998** [comment]
I get two different results with the same set of inputs and then calculate the cosine distance (similarity) with the following code
           text1 = ["mov rbp rdi",
                "mov ebx 0x1",
                "mov rdx rbx",
                "call memcpy",
                "mov [ rcx + rbx ] 0x0",
                "mov rcx rax",
                "mov [ rax ] 0x2e"]
        text2 = ["mov rbp rdi",
                 "mov ebx 0x1",
                 "mov rdx rbx",
                 "call memcpy",
                 "mov [ rcx + rbx ] 0x0",
                 "mov rcx rax",
                 "mov [ rax ] 0x2e"]
        embeddings1 = palmtree.encode(text1)
        embeddings2 = palmtree.encode(text2)
        x1 = embeddings1 / norm(embeddings1, axis=-1, keepdims=True)
        x2 = embeddings2 / norm(embeddings2, axis=-1, keepdims=True)
        cos1 = np.dot(x1, x2.T)
        x11 = normalize(embeddings1)
        x22 = normalize(embeddings2)
        cos2 = np.dot(x11, x22.T)
        np.testing.assert_allclose(cos1, cos2, rtol=1e-6,atol=0)
No error is reported when the parameter is 1e-6, and an error is reported when adjusted to 1e-7
It can be considered that although the data looks different, its position in space is same

**Comment by Alex-Ruan** [comment]
> Not an expected result. Need to check the code.

this is maybe the reason of random initial, I fix the random seed, then I get the same result everytime ``

**Comment by Alex-Ruan** [comment]
> 

this maybe because the random initialization, I fix the random seed, then I get the same embeddings everytime I run the how2use.py`def set_seed(seed=42):
    random.seed(seed)
    os.environ['PYHTONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True`

**Comment by Alex-Ruan** [comment]
> Not an expected result. Need to check the code.

this maybe because the random initialization, I fix the random seed, then I get the same embeddings everytime I run the how2use.py`def set_seed(seed=42): random.seed(seed) os.environ['PYHTONHASHSEED'] = str(seed) np.random.seed(seed) torch.manual_seed(seed) torch.cuda.manual_seed(seed) torch.backends.cudnn.deterministic = True`

**Comment by Alex-Ruan** [comment]
Hi，邮件已收到，谢谢。

**Comment by thibaultrichel** [comment]
As @moritzdannehl said, the `eval()` method from torch models is working like a switch for some layers, as explained here : https://stackoverflow.com/questions/60018578/what-does-model-eval-do-in-pytorch

So calling `self.model.eval()` in the `UsableTransformer` class in `eval_utils.py` just after calling `self.model = torch.load(model_path)` will do the trick.

**Comment by SidekickLx** [comment]
Many thanks to  @moritzdannehl  and @thibaultrichel 
I've updated how2use.py and included it in our repo.

**Comment by Alex-Ruan** [comment]
Hi，邮件已收到，谢谢。
